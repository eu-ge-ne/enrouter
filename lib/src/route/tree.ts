import type { Route } from "./mod.js";
//@ts-ignore
import routeTree from "virtual:enrouter";

export function getRouteTree(): Route {
  return routeTree;
}
