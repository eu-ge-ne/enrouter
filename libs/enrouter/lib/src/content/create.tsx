import type { ReactElement } from "react";

import type { Match } from "#lib/match/mod.js";
import type { Content } from "./mod.js";
import { logger } from "#lib/debug.js";
import { ContentProvider } from "./context.js";
import { NotFound } from "./notFound.js";

const log = logger("content/create");

export function createContent(matches: Match[]): ReactElement[] {
  log("Creating matched content");

  const content = matches.map(toContent);
  const elements = render(content);

  log("Matched content created");

  return elements;
}

function toContent(match: Match): Content {
  const content: Content = {
    match,
  };

  function wrapChildren([key, children]: [string, ReactElement]) {
    const el = (
      <ContentProvider key={key} value={content}>
        {children}
      </ContentProvider>
    );
    return [key, el];
  }

  // do not render if match.next !== undefined
  if (!match.next && match.route.elements.index) {
    content.index = Object.fromEntries(
      Object.entries(match.route.elements.index).map((entry) => {
        log(
          'Creating content from index component "%s" matched by "%s"',
          entry[0],
          match.route.path,
        );
        return wrapChildren(entry);
      }),
    );
  }

  if (match.route.elements.layout) {
    content.layout = Object.fromEntries(
      Object.entries(match.route.elements.layout).map((entry) => {
        log(
          'Creating content from layout component "%s" matched by "%s"',
          entry[0],
          match.route.path,
        );
        return wrapChildren(entry);
      }),
    );
  }

  return content;
}

function render(content: Content[]): ReactElement[] {
  // 404?
  if (!content.at(-1)?.match.isFull) {
    log("404: %o", content);

    const i = content.findLastIndex((x) => x.match.route.elements.notFound);
    if (i === -1) {
      return [<NotFound />];
    }

    content = content.slice(0, i + 1);

    const x = content[i]!;
    x.index = x.match.route.elements.notFound;
  }

  const last = content.at(-1);
  content.forEach((x, i) => {
    x.prev = content[i - 1];
    x.next = content[i + 1];
    x.last = last;
  });

  return Object.values(content[0]?.layout ?? {});
}
