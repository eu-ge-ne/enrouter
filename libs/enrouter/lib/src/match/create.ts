import type { Route } from "#lib/route/mod.js";
import type { Match } from "./mod.js";
import { logger } from "#lib/debug.js";
import { loadRoutes } from "#lib/route/load.js";

const log = logger("match/create");

export interface CreateMatchParams {
  routes: Route;
  location: string;
}

export async function createMatch({
  routes,
  location,
}: CreateMatchParams): Promise<Match | undefined> {
  const matches: Match[] = [];

  recur([routes], location, matches);

  await loadRoutes(matches.map((x) => x.route));

  trim(matches);
  link(matches);

  log("matched: %o", matches);

  return matches[0];
}

function recur(routes: Route[], location: string, matches: Match[]): void {
  const match = matchOneOf(routes, location);
  if (!match) {
    return;
  }

  log(`path: "%s", location: "%s"`, match.route.path, match.location);

  matches.push(match);

  if (match.route.tree) {
    recur(match.route.tree, location, matches);
  }
}

function matchOneOf(routes: Route[], location: string): Match | undefined {
  let matched = routes
    .map((route) => {
      const execs = route.test.pattern.exec(location);
      if (execs) {
        return { route, execs };
      }
    })
    .filter((x) => x !== undefined);

  if (matched.length === 0) {
    return;
  }

  // TODO: improve

  if (matched.length > 1) {
    matched = matched.filter((x) => x.route.test.keys.length === 0);
  }

  if (matched.length !== 1) {
    throw new Error("Found >1 match");
  }

  const { route, execs } = matched[0]!;
  const loc = execs[0] || "/";

  return {
    route,
    isFull: loc === location,
    location: loc,
    params: Object.fromEntries(
      route.test.keys.map((key, i) => [key, execs![i + 1]!] as const),
    ),
  };
}

function trim(matches: Match[]): void {
  const rootIndex = matches.findLastIndex((x) => x.route.elements._root);
  if (rootIndex >= 0) {
    matches.splice(0, rootIndex);
  }

  if (!matches.at(-1)?.isFull) {
    const voidIndex = matches.findLastIndex(
      ({
        route: {
          elements: { _void, __void },
        },
      }) => _void || __void,
    );

    if (voidIndex >= 0) {
      if (matches[voidIndex]?.route.elements.__void) {
        const match = matches[voidIndex];
        matches.splice(0, matches.length);
        matches.push(match);
      } else {
        matches.splice(voidIndex + 1);
      }
    }
  }
}

function link(matches: Match[]): void {
  const first = matches[0];
  const last = matches.at(-1);

  matches.forEach((x, i) => {
    x.first = first;
    x.next = matches[i + 1];
    x.last = last;
  });
}
