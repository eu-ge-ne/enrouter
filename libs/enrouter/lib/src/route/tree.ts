import type { Route } from "./mod.js";
//@ts-ignore
import routes from "virtual:enrouter/vite/routes";

export const routeTree = routes as Route;
