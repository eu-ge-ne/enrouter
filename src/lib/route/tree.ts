import type { Route } from "./route.js";
//@ts-ignore
import routeTree from "virtual:enrouter";

export function getRouteTree(): Route {
  return routeTree;
}
