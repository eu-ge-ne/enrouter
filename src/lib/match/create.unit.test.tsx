import { describe, test, expect, vi } from "vitest";
import * as regexparam from "regexparam";

import { getRouteTree } from "#lib/route/tree.js";
import { matchLocation } from "./create.js";

vi.mock(import("#lib/route/tree.js"), () => ({
  getRouteTree: vi.fn(),
}));

describe("match", () => {
  describe("matchLocation", () => {
    test("0 matches", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/abc",
        test: regexparam.parse("/abc", true),
        modules: [],
        loaded: false,
        elements: {},
      });

      expect(await matchLocation("/x")).toMatchSnapshot();
    });

    test("from 1 route", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: false,
        elements: {},
      });

      expect(await matchLocation("/")).toMatchSnapshot();
    });

    test("from 2 routes", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: false,
        elements: {},
        tree: [
          {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: false,
            elements: {},
          },
        ],
      });

      expect(await matchLocation("/abc")).toMatchSnapshot();
    });

    test("from 1 route with params", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/[:id]",
        test: regexparam.parse("/:id", true),
        modules: [],
        loaded: false,
        elements: {},
      });

      expect(await matchLocation("/100")).toMatchSnapshot();
    });

    test("from 2 routes with params", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: false,
        elements: {},
        tree: [
          {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: false,
            elements: {},
          },
          {
            path: "/[:id]",
            test: regexparam.parse("/:id", true),
            modules: [],
            loaded: false,
            elements: {},
          },
        ],
      });

      expect(await matchLocation("/abc")).toMatchSnapshot();
    });

    test('from 3 routes with "_void"', async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: true,
        elements: {},
        tree: [
          {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: true,
            elements: {
              _void: {
                Main: <div>Void</div>,
              },
            },
            tree: [
              {
                path: "/abc/xyz",
                test: regexparam.parse("/abc/xyz", true),
                modules: [],
                loaded: true,
                elements: {},
              },
            ],
          },
        ],
      });

      expect(await matchLocation("/abc/xyz/100")).toMatchSnapshot();
    });
  });
});
