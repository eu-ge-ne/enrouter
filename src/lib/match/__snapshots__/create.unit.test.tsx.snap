// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`match > matchLocation > 0 matches 1`] = `[]`;

exports[`match > matchLocation > from 1 route 1`] = `
[
  {
    "isExact": true,
    "isVoid": false,
    "location": "/",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > matchLocation > from 1 route with params 1`] = `
[
  {
    "isExact": true,
    "isVoid": false,
    "location": "/100",
    "params": {
      "id": "100",
    },
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/[:id]",
      "test": {
        "keys": [
          "id",
        ],
        "pattern": /\\^\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > matchLocation > from 2 routes 1`] = `
[
  {
    "isExact": false,
    "isVoid": false,
    "location": "/",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {},
          "loaded": false,
          "modules": [],
          "path": "/abc",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
      ],
    },
  },
  {
    "isExact": true,
    "isVoid": false,
    "location": "/abc",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/abc",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > matchLocation > from 2 routes with params 1`] = `
[
  {
    "isExact": false,
    "isVoid": false,
    "location": "/",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {},
          "loaded": false,
          "modules": [],
          "path": "/abc",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
        {
          "elements": {},
          "loaded": false,
          "modules": [],
          "path": "/[:id]",
          "test": {
            "keys": [
              "id",
            ],
            "pattern": /\\^\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
      ],
    },
  },
  {
    "isExact": true,
    "isVoid": false,
    "location": "/abc",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/abc",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > matchLocation > from 3 routes with "_void" 1`] = `
[
  {
    "isExact": false,
    "isVoid": false,
    "location": "/",
    "params": {},
    "route": {
      "elements": {},
      "loaded": true,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {
            "_void": {
              "Main": <div>
                Void
              </div>,
            },
          },
          "loaded": true,
          "modules": [],
          "path": "/abc",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
          },
          "tree": [
            {
              "elements": {},
              "loaded": true,
              "modules": [],
              "path": "/abc/xyz",
              "test": {
                "keys": [],
                "pattern": /\\^\\\\/abc\\\\/xyz\\(\\?=\\$\\|\\\\/\\)/i,
              },
            },
          ],
        },
      ],
    },
  },
  {
    "isExact": false,
    "isVoid": true,
    "location": "/abc",
    "params": {},
    "route": {
      "elements": {
        "_void": {
          "Main": <div>
            Void
          </div>,
        },
      },
      "loaded": true,
      "modules": [],
      "path": "/abc",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {},
          "loaded": true,
          "modules": [],
          "path": "/abc/xyz",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\\\/xyz\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
      ],
    },
  },
  {
    "isExact": false,
    "isVoid": false,
    "location": "/abc/xyz",
    "params": {},
    "route": {
      "elements": {},
      "loaded": true,
      "modules": [],
      "path": "/abc/xyz",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\\\/xyz\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;
